HeartbeatLogging/Sources/HeartbeatsBundle.swift

```diff
@@ -72,6 +72,30 @@ struct HeartbeatsBundle: Codable, HeartbeatsPayloadConvertible {
     }
   }
 
+  /// Removes the heartbeat associated with the given date.
+  /// - Parameter date: The date of the heartbeat needing removal.
+  /// - Returns: The heartbeat that was removed or `nil` if there was no heartbeat to remove.
+  @discardableResult
+  mutating func removeHeartbeat(from date: Date) -> Heartbeat? {
+    var removedHeartbeat: Heartbeat?
+
+    var poppedHeartbeats: [Heartbeat] = []
+
+    while let poppedHeartbeat = buffer.pop() {
```

Comment:
Why a stack data structure that requires popping and repushing to extract versus a list in which extracting would be simpler?

Reply:
Good question. 2 factors informed this decision. 

- 1. The underlying heartbeat data structure is a ring buffer* and it internally manages a fixed capacity array and index (`tailIndex`) where it pushes/pops from. The buffer does conform to Swift's `Sequence` so we could filter the heartbeat out but then we are left with an array that is has one less capacity than the buffer it came from and we arent sure where the index index should be. So a Ringbuffer initializer would need to be added that takes in a list, a capacity, and a tail index and we'd have to rebuild the data structure through that. 

- 2. Because of reason one, I figured it'd be easier to add an API to the RingBuffer type. I added `pop` API because it is another API you would expect a ring buffer implementation to have. With `push` and `pop` you can remove elements from the buffer without risking the insertion index being compromised.

FWIW, 99% of the time, the last heartbeat in the buffer will be the heartbeat from today so this logic will take the happiest path (meaning first heartbeat popped is the heartbeat from today). The reason it is written this way is to handle the 1% edge case where the device's system time has been modified by end user and the expected ordering breaks down. Because of that edge case, we cant make assumption that last heartbeat inserted is heartbeat from today. 

It's very important that the heartbeat we ultimately remove is from today since the date will be assigned by the backend (this is all for v1 clients that send FIRHeartbeatInfo codes.)

> *Ring buffer was chosen since we want to store a fixed capacity of heartbeats where the oldest heartbeat is overwritten when the capacity is reached. I thought it worked well for this use case since we append based on date. Plus it has constant time operations.